src/solvers/refinement/string_constraint_generator_valueof.cpp

```
@@ -574,40 +608,148 @@ exprt is_digit_with_radix(exprt chr, exprt radix_char_type)
+  }
+}
+
+/// Calculate the string length needed to represent any value of the given type
+/// using the given radix
+/// \param type: the type that we are considering values of
+/// \param radix: the radix we are using, as a double
+/// \return the maximum string length
+size_t calculate_max_string_length(const typet &type, double radix)
+{
+  double n_bits=static_cast<double>(to_bitvector_type(type).get_width());
+
+  /// If the type is signed then the maximum absolute value that needs to be
+  /// representable is one bit smaller
+  bool signed_type=type.id()==ID_signedbv;
+  /// Let m be the minimal number of characters needed.
+  /// For signed types, the longest string will be for -2^(n-1), so
+  ///   m = 1 + min{k: 2^(n-1) < r^k}
+  ///     = 1 + min{k: n-1 < k log_2(r)}
+  ///     = 1 + min{k: k > (n-1) / log_2(r)}
+  ///     = 1 + min{k: k > floor((n-1) / log_2(r))}
+  ///     = 1 + (1 + floor((n-1) / log_2(r)))
+  ///     = 2 + floor((n-1) / log_2(r))
+  /// For unsigned types, the longest string will be for (2^n)-1, so
+  ///   m = min{k: (2^n)-1 < r^k}
+  ///     = min{k: 2^n <= r^k}
+  ///     = min{k: n <= k log_2(r)}
+  ///     = min{k: k >= n / log_2(r)}
+  ///     = min{k: k >= ceil(n / log_2(r))}
+  ///     = ceil(n / log_2(r))
+  double max_string_length=signed_type?
+    floor(static_cast<double>(n_bits-1)/log2(radix))+2.0:
+    ceil(static_cast<double>(n_bits)/log2(radix));
+  return static_cast<size_t>(max_string_length);
+}
+
+/// If the expression is a constant expression then we get the value of it as
+/// an unsigned long. If not we return a default value.
+/// \param expr: input expression
+/// \param def: default value to return if we cannot evaluate expr
+/// \return the output as an unsigned long
+unsigned long try_to_evaluate_expr_as_ul(const exprt &expr, unsigned long def)
+{
+  symbol_tablet symtab;
+  const namespacet ns(symtab);
+  mp_integer mp_radix;
+  bool to_integer_failed=to_integer(simplify_expr(expr, ns), mp_radix);
```

@smowton explained to me that `simplify_expr` will crash if it finds a symbol that isn't in the namespace, rather than giving up as I had assumed. I will try to get access to the namespace. If that isn't possible, I'll have to give up on evaluating the radix, which is only for optimisation purposes.